File: 20/python/aoc4.py
# part 1
# def is_passport_valid(passport):
#     valid_fields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
#     for field in valid_fields:
#         if field not in passport:
#             return False
#     return True


# def passport(data):
#     current_passport = {}
#     valid = 0
#     for line in data:
#         line = line.strip()
#         # time to check if the passport is valid
#         if len(line) < 1:
#             if is_passport_valid(current_passport):
#                 valid += 1
#             current_passport = {}
#         else:
#             split = line.split(" ")
#             for fields in split:
#                 field = fields.split(":")
#                 current_passport[field[0]] = field[1]
#     if is_passport_valid(current_passport):
#         valid += 1

#     return valid



# if __name__ == "__main__":
#     text_file = open("../input/aoc4.txt", "r")
#     string_data = text_file.readlines()
#     ans = passport(string_data)
#     print("the number of valid passports is {}".format(ans))

import re

def is_passport_valid(passport):
    items = passport.items()
    valid_fields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
    for field in valid_fields:
        if field not in passport:
            return False
    for item in items:
        field = item[0]
        val = item[1]
        if field == "byr":
            if int(val) > 2002 or int(val) < 1920:
                return False
        elif field == "iyr":
            if int(val) > 2020 or int(val) < 2010:
                return False
        elif field == "eyr":
            if int(val) > 2030 or int(val) < 2020:
                return False
        elif field == "hgt":
            if val[-2:] == "cm":
                if int(val[:-2]) > 193 or int(val[:-2]) < 150:
                    return False
            elif val[-2:] == "in":
                if int(val[:-2]) > 76 or int(val[:-2]) < 59:
                        return False
        elif field == "hcl":
            if len(val) == 7:
                if not re.search("#([a-f0-9])", val):
                    return False
            else:
                return False
        elif field == "ecl":
            valid_colors = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
            if val in valid_colors:
                continue
            else:
                return False
        elif field == "pid":
            if len(val) != 9:
                return False
    return True


def passport(data):
    current_passport = {}
    valid = 0
    for line in data:
        line = line.strip()
        # time to check if the passport is valid
        if len(line) < 1:
            if is_passport_valid(current_passport):
                valid += 1
            current_passport = {}
        else:
            split = line.split(" ")
            for fields in split:
                field = fields.split(":")
                current_passport[field[0]] = field[1]
    if is_passport_valid(current_passport):
        valid += 1
        
    return valid



if __name__ == "__main__":
    text_file = open("../input/aoc4.txt", "r")
    string_data = text_file.readlines()
    ans = passport(string_data)
    print("the number of valid passports is {}".format(ans))

File: 20/go/aoc4.go
package main

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	data, err := ReadFile("../input/aoc4.txt")
	if err != nil {
		fmt.Printf("readFile failed: %v\n", err)
	}

	ans := validPassports(data)

	fmt.Printf("The number of valid passports is %d \n", ans)
}

func validPassports(data []string) int {
	valid := 0
	currentPassport := make(map[string]string)
	for _, line := range data {
		if len(line) <= 1 {
			// we are at an empty line and we need to clear the current passport
			if isPassportValid2(currentPassport) {
				valid++
			}
			currentPassport = make(map[string]string)
		} else {
			splitBySpace := strings.Split(line, " ")
			for _, kv := range splitBySpace {
				splitByColon := strings.Split(kv, ":")
				currentPassport[splitByColon[0]] = splitByColon[1]
			}
		}
	}
	if len(currentPassport) > 1 {
		if isPassportValid2(currentPassport) {
			valid++
		}
	}
	return valid
}

func isPassportValid(passport map[string]string) bool {
	validFields := []string{"byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"}
	valid := true
	for _, field := range validFields {
		if _, ok := passport[field]; !ok {
			valid = false
		}
	}
	return valid
}

func isPassportValid2(passport map[string]string) bool {
	validFields := []string{"byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"}
	valid := true
	for _, field := range validFields {
		if val, ok := passport[field]; ok {
			switch field {
			case "byr":
				valid = checkIntVal(val, 1920, 2002)
			case "iyr":
				valid = checkIntVal(val, 2010, 2020)
			case "eyr":
				valid = checkIntVal(val, 2020, 2030)
			case "hgt":
				valid = checkHgt(val)
			case "hcl":
				valid = checkHcl(val)
			case "ecl":
				valid = checkEcl(val)
			case "pid":
				valid = checkPid(val)
			default:
			}
		} else {
			// key isn't in the passport
			valid = false
		}
		if !valid {
			// if we found an invalid value, just break the loop
			break
		}
	}
	if valid {
		fmt.Println("valid: ", passport)
	}
	return valid
}

func checkIntVal(val string, low int, high int) bool {
	valid := true
	intval, err := strconv.Atoi(val)
	if err != nil {
		valid = false
	}
	if intval < low || intval > high {
		valid = false
	}
	return valid
}

func checkHgt(val string) bool {
	valid := true
	if strings.Contains(val, "in") {
		idx := strings.Index(val, "in")
		hgt := val[:idx]
		valid = checkIntVal(hgt, 59, 76)
	} else if strings.Contains(val, "cm") {
		idx := strings.Index(val, "cm")
		hgt := val[:idx]
		valid = checkIntVal(hgt, 150, 193)
	} else {
		valid = false
	}
	return valid
}

func checkHcl(val string) bool {
	match, _ := regexp.MatchString("#([a-f0-9])", val)
	return match
}

func checkEcl(val string) bool {
	valid := false
	validColors := []string{"amb", "blu", "brn", "gry", "grn", "hzl", "oth"}
	for _, col := range validColors {
		if val == col {
			valid = true
		}
	}
	return valid
}

func checkPid(val string) bool {
	valid := true
	if len(val) != 9 {
		valid = false
	}
	_, err := strconv.Atoi(val)
	if err != nil {
		valid = false
	}
	return valid
}
